File: 2014:2015/StackEventsExceptions/CppStacks/cstack.cpp
#include "cstack.h"

cstack::cstack(int c)
{
    cap = c;
}

void cstack::push(string element){
    //cout << stack.capacity() << " " << stack.size() << endl;
    if(cap < stack.size() + 1){
        throw new StackOverflowException("Reached capacity of vector.");
    }else{
        stack.push_back(element);
    }
}

string cstack::pop(){
    if(stack.size() == 0){
        OnNoElements();
    }else{
        return stack.back();
    }
    return "";
}

void cstack::clear(){
    stack.clear();
}

File: 2014:2015/StackEventsExceptions/CSStackEventsException/CSStackEventsException/CStack.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace CSStackEventsException
{
	public delegate void StackDelegate();

	public class CStack
	{
		List<String> Stack;

		public int capacity{ get; private set; }

		public CStack (int cap /*capacity*/)
		{
			capacity = cap;
			Stack = new List<String> (cap);
		}

		public void push(string element){
			// Abfrage: Kapazitaet erreicht?
			// ja: -> Exception ausloesen
			// nein: -> Element eintragen
			if(Stack.Capacity < Stack.Count + 1){
				throw new CStackOverflowException ();
			}else{
				Stack.Add (element);
			}
		}

		public string pop(){
			// Abfrage: noch ein Element in der Liste vorhanden?
			// ja: -> Element returnen
			// nein: -> Event ausloesen

			if(Stack.Count != 0){
				return Stack.Last();
			}else{
				underFlow ();
			}

			return string.Empty;
		}

		public void clear(){
			Stack.Clear ();
		}

		public event StackDelegate underFlow;
	}

	class CStackOverflowException:ApplicationException{
		// -> 3 Konstruktoren!

		public CStackOverflowException(){
		}

		public CStackOverflowException(string message):base(message){
		}

		public CStackOverflowException(string message, Exception inner):base(message, inner){
		}
	}
}

